How to create NFT using Contract Tools tutorial (#2874)
near/docs: docs/primitives/nft/sdk-contract-tools.md

---
id: nft-contract-tools
title: Create NFT using Contract Tools
sidebar_label: Create NFT using Contract Tools
description: "Learn how to create a non-fungible token (NFT) using Contract Tools package"
---

import {Github} from "@site/src/components/UI/Codetabs";
import MovingForwardSupportSection from '@site/src/components/MovingForwardSupportSection';

In this tutorial, we will create a non-fungible token (NFT) using the [NEAR SDK Contract Tools](https://github.com/near/near-sdk-contract-tools) package. This package is a collection of common tools and patterns to simplify smart contract development, including:

- Storage fee management
- Escrow pattern and derive macro
- Owner pattern and derive macro
- Pause pattern and derive macro
- Role-based access control
- Derive macros for [NEP standards](./standard.md)
  - NEP-141 (fungible token), extension NEP-148
  - NEP-145 (storage management), and integrations for the fungible token and non-fungible token standards
  - NEP-171 (non-fungible token), extensions NEP-177, NEP-178, NEP-181

>   - NEP-297 (events)

Reviewer: ```suggestion
  - [NEP-141](https://github.com/near/NEPs/blob/master/neps/nep-0141.md) (fungible token), extension [NEP-148](https://github.com/near/NEPs/blob/master/neps/nep-0148.md)
  - [NEP-145](https://github.com/near/NEPs/blob/master/neps/nep-0145.md) (storage management), and integrations for the fungible token and non-fungible token standards
  - [NEP-171](https://github.com/near/NEPs/blob/master/neps/nep-0171.md) (non-fungible token), extensions [NEP-177](https://github.com/near/NEPs/blob/master/neps/nep-0177.md), [NEP-178](https://github.com/near/NEPs/blob/master/neps/nep-0178.md), [NEP-181](https://github.com/near/NEPs/blob/master/neps/nep-0181.md)
  - [NEP-297](https://github.com/near/NEPs/blob/master/neps/nep-0297.md) (events)
```

---

## Introduction

While one can create a non-fungible token (NFT) contract from scratch using only the `near-sdk` and `near_contract_standards` (e.g. [NFT contract](https://github.com/near-examples/NFT)), a simpler approach is to use `near-sdk-contract-tools`.

`near-sdk-contract-tools` allows us implement the logic for minting/burning logic, access control, and other NFT standards by simply deriving macros on our contract struct, as `OpenZeppelin` does for Ethereum contracts.

---

## Basic NFT Methods

> To derive basic NFT methods to our contract, we need to derive `NonFungibleToken` macro to our contract struct:

Reviewer: ```suggestion
To derive basic NFT methods to your smart contract, you need to derive the `NonFungibleToken` macro to the contract's `struct`:
```
